src/libbluechi/cli/command.c

```diff
@@ -0,0 +1,141 @@
+        }
+
+        command->ref_count = 1;
+        command->op = NULL;
+        command->opargc = 0;
+        command->opargv = NULL;
+        command->is_help = false;
+
+        LIST_HEAD_INIT(command->command_options);
+
+        return steal_pointer(&command);
+}
+
+Command *command_ref(Command *command) {
+        command->ref_count++;
+        return command;
+}
+
+void command_unref(Command *command) {
+        command->ref_count--;
+        if (command->ref_count != 0) {
+                return;
+        }
+
+        CommandOption *option = NULL;
+        CommandOption *next_option = NULL;
+        LIST_FOREACH_SAFE(options, option, next_option, command->command_options) {
+                command_option_free(option);
+        }
+
+        free_and_null(command);
+}
+
+int command_execute(Command *command, void *userdata) {
+        const Method *method = command->method;
+        if (command->is_help) {
+                method->usage();
+        }
+        if (command->opargc < method->min_args) {
+                fprintf(stderr, "Too few arguments for method %s\n", method->name);
+                return -EINVAL;
+        }
+        if (command->opargc > method->max_args) {
+                fprintf(stderr, "Too many arguments for method %s\n", method->name);
+                return -EINVAL;
+        }
+        CommandOption *opt = NULL;
+        LIST_FOREACH(options, opt, command->command_options) {
+                if (!(opt->option_type->option_flag & method->supported_options)) {
+                        fprintf(stderr,
+                                "Unsupported option '%s' for method '%s'\n",
+                                opt->option_type->option_long,
+                                method->name);
+                        return -EINVAL;
+                }
+        }
+
+        return method->dispatch(command, userdata);
+}
+
+void command_option_free(CommandOption *option) {
+        free_and_null(option->value);
+        free_and_null(option);
+}
+
+char *command_get_option(Command *command, int key) {
+        CommandOption *option = NULL;
+        LIST_FOREACH(options, option, command->command_options) {
+                if (key == option->key) {
+                        return option->value;
+                }
+        }
+        return NULL;
+}
+
+bool command_check_flag(Command *command, int key) {
+        CommandOption *option = NULL;
+        LIST_FOREACH(options, option, command->command_options) {
+                if (key == option->key) {
+                        return true;
+                }
+        }
+        return false;
+}
+
+int command_add_option(Command *command, int key, char *value, const OptionType *option_type) {
+        CommandOption *opt = malloc0(sizeof(CommandOption));
+        if (opt == NULL) {
+                return -ENOMEM;
+        }
+
+        opt->key = key;
+        if (value != NULL) {
+                opt->is_flag = false;
+                opt->value = strdup(value);
```

Comment:
`strdup` allocates memory. Should we check that it didn't return `NULL`?

Reply:
Yes. And also added cleanup for `CommandOption` for that case